config/v1/types_oauth.go

```diff
@@ -1,37 +1,478 @@
 package v1
 
-import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OAuth Server and Identity Provider Config
 
 // +genclient
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // OAuth holds cluster-wide information about OAuth.  The canonical name is `cluster`
-// TODO this object is an example of a possible grouping and is subject to change or removal
 type OAuth struct {
-	metav1.TypeMeta `json:",inline"`
-	// Standard object's metadata.
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	// spec holds user settable values for configuration
-	Spec OAuthSpec `json:"spec"`
-	// status holds observed values from the cluster. They may not be overridden.
-	Status OAuthStatus `json:"status"`
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              OAuthSpec   `json:"spec"`
+	Status            OAuthStatus `json:"status,omitempty"`
 }
 
 type OAuthSpec struct {
-	// options for configuring the embedded oauth server.
-	// possibly wellknown?
+	//IdentityProviders is an ordered list of ways for a user to identify themselves
+	IdentityProviders []OAuthIdentityProvider `json:"identityProviders"`
+
+	// TokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// Templates allow you to customize pages like the login page.
+	// +optional
+	Templates OAuthTemplates `json:"templates"`
 }
 
 type OAuthStatus struct {
+	// TODO Fill in
 }
 
-// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+const (
+	// LoginTemplateKey is the key of the login template
+	LoginTemplateKey = "login.html"
+	// ProviderSelectionTemplateKey is the key for the provider selection template
+	ProviderSelectionTemplateKey = "providers.html"
+	// ErrorsTemplateKey is the key for the errors template
+	ErrorsTemplateKey = "errors.html"
+)
+
+// OAuthTemplates allow for customization of pages like the login page
+type OAuthTemplates struct {
+	// Login is a reference to a secret that specifies a go template to use to render the login page.
+	// The secret referenced must contain a key named `login.html` containing the template data.
+	// If unspecified, the default login page is used.
+	// +optional
+	Login corev1.SecretReference `json:"login"`
```

Comment:
Define a peer to `ConfigMapReference` in this package.  We don't allow namespace choices.

Reply:
Is there a reason to create a new type instead of the ones in core? It seems like [`corev1.SecretKeySelector`](https://godoc.org/k8s.io/api/core/v1#SecretKeySelector) is exactly what we'd want.